File: 04-batch-performance/ai/java/src/main/java/com/treinamento/batch/datastructures/LinkedList.java
package com.treinamento.batch.datastructures;

import java.util.Comparator;
import java.util.NoSuchElementException;
import java.util.function.Predicate;

/**
 * Custom singly-linked list.
 *
 * <p><strong>Intentional performance characteristics:</strong>
 * Every method that searches or inserts by position is O(n). There is no
 * index, no hash table, and no skip structure — just a plain chain of
 * {@link Node} objects traversed from head to tail on every operation.
 *
 * <p>This is the <em>correct</em> but <em>slow</em> implementation that
 * the exercise asks you to replace with an efficient alternative.
 *
 * @param <T> type of elements stored in this list
 */
public class LinkedList<T> {

    // -------------------------------------------------------------------------
    // Inner node class
    // -------------------------------------------------------------------------

    /** A single link in the chain. */
    public static class Node<T> {
        public T data;
        public Node<T> next;

        public Node(T data) {
            this.data = data;
            this.next = null;
        }
    }

    // -------------------------------------------------------------------------
    // State
    // -------------------------------------------------------------------------

    private Node<T> head;
    private int size;

    // -------------------------------------------------------------------------
    // Constructors
    // -------------------------------------------------------------------------

    /** Creates an empty list. */
    public LinkedList() {
        this.head = null;
        this.size = 0;
    }

    // -------------------------------------------------------------------------
    // Core operations — all O(n) by design
    // -------------------------------------------------------------------------

    /**
     * Appends {@code element} at the tail of the list.
     *
     * <p>O(n) — walks the entire chain to reach the last node.
     * An efficient implementation would keep a {@code tail} pointer and make
     * this O(1).
     */
    public void append(T element) {
        Node<T> newNode = new Node<>(element);
        if (head == null) {
            head = newNode;
        } else {
            Node<T> current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = newNode;
        }
        size++;
    }

    /**
     * Returns the first element matching {@code predicate}, or {@code null}
     * if none is found.
     *
     * <p>O(n) — unavoidable for an unsorted list without an index.
     */
    public T find(Predicate<T> predicate) {
        Node<T> current = head;
        while (current != null) {
            if (predicate.test(current.data)) {
                return current.data;
            }
            current = current.next;
        }
        return null;
    }

    /**
     * Returns all elements as an array.
     *
     * <p>O(n) — one pass to collect elements. The array type is
     * {@code Object[]} because Java generics do not allow {@code new T[size]}.
     * Callers must cast the result.
     */
    @SuppressWarnings("unchecked")
    public T[] toArray() {
        Object[] arr = new Object[size];
        Node<T> current = head;
        int i = 0;
        while (current != null) {
            arr[i++] = current.data;
            current = current.next;
        }
        return (T[]) arr;
    }

    /**
     * Inserts {@code element} into the list so that the list remains sorted
     * according to {@code comparator}.
     *
     * <p>O(n) — must scan from head to find the insertion point. Combined
     * with repeated calls this produces <strong>O(n²)</strong> behaviour for
     * building a fully-sorted list (insertion sort).
     */
    public void insertSorted(T element, Comparator<T> comparator) {
        Node<T> newNode = new Node<>(element);

        // Insert before head if list is empty or element is the smallest
        if (head == null || comparator.compare(element, head.data) <= 0) {
            newNode.next = head;
            head = newNode;
            size++;
            return;
        }

        Node<T> current = head;
        while (current.next != null
                && comparator.compare(element, current.next.data) > 0) {
            current = current.next;
        }
        newNode.next = current.next;
        current.next = newNode;
        size++;
    }

    /**
     * Removes the first element matching {@code predicate}.
     *
     * <p>O(n) — linear scan.
     *
     * @return {@code true} if an element was removed, {@code false} otherwise
     */
    public boolean remove(Predicate<T> predicate) {
        if (head == null) return false;

        if (predicate.test(head.data)) {
            head = head.next;
            size--;
            return true;
        }

        Node<T> current = head;
        while (current.next != null) {
            if (predicate.test(current.next.data)) {
                current.next = current.next.next;
                size--;
                return true;
            }
            current = current.next;
        }
        return false;
    }

    /**
     * Returns the first element of the list without removing it.
     *
     * @throws NoSuchElementException if the list is empty
     */
    public T peek() {
        if (head == null) throw new NoSuchElementException("List is empty");
        return head.data;
    }

    /**
     * Removes and returns the first element of the list.
     *
     * @throws NoSuchElementException if the list is empty
     */
    public T poll() {
        if (head == null) throw new NoSuchElementException("List is empty");
        T data = head.data;
        head = head.next;
        size--;
        return data;
    }

    /** Returns the number of elements in this list. O(1). */
    public int size() {
        return size;
    }

    /** Returns {@code true} if this list contains no elements. */
    public boolean isEmpty() {
        return size == 0;
    }

    /** Provides access to the head node for iteration. */
    public Node<T> getHead() {
        return head;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("[");
        Node<T> current = head;
        while (current != null) {
            sb.append(current.data);
            if (current.next != null) sb.append(", ");
            current = current.next;
        }
        sb.append("]");
        return sb.toString();
    }
}

File: 04-batch-performance/ai/kotlin/src/main/kotlin/com/treinamento/batch/datastructures/LinkedList.kt
package com.treinamento.batch.datastructures

/**
 * Singly-linked list node.
 *
 * Kept intentionally public so [LinkedList] and external helpers can
 * traverse the chain directly without going through the list API.
 */
class Node<T>(var value: T, var next: Node<T>? = null)

/**
 * Custom singly-linked list.
 *
 * INTENTIONAL PERFORMANCE NOTE
 * ─────────────────────────────
 * Every operation here is O(n):
 *   - [find]         – full sequential scan
 *   - [insertSorted] – full sequential scan to locate insertion point
 *   - [size]         – traverses every node on each call (no cached counter)
 *   - [toList]       – O(n) traversal
 *   - [append]       – O(n) walk to tail (no tail pointer kept)
 *
 * These properties are exploited in [com.treinamento.batch.BatchProcessor]
 * to produce measurable O(n²) bottlenecks.
 */
class LinkedList<T> {

    /** Head of the chain; null when the list is empty. */
    var head: Node<T>? = null
        private set

    // ── Basic operations ──────────────────────────────────────────────────

    /**
     * Appends [value] to the tail of the list.
     * O(n) – walks the entire chain to find the last node.
     */
    fun append(value: T) {
        val newNode = Node(value)
        val current = head
        if (current == null) {
            head = newNode
            return
        }
        // Walk all the way to the end – O(n) intentionally (no tail pointer).
        var tail: Node<T> = current
        while (tail.next != null) {
            tail = tail.next!!
        }
        tail.next = newNode
    }

    /**
     * Returns the first element that satisfies [predicate], or null.
     * O(n) – sequential scan of every node.
     */
    fun find(predicate: (T) -> Boolean): T? {
        var current = head
        while (current != null) {
            if (predicate(current.value)) return current.value
            current = current.next
        }
        return null
    }

    /**
     * Materialises the list as a [kotlin.collections.List].
     * O(n).
     */
    fun toList(): List<T> {
        val result = mutableListOf<T>()
        var current = head
        while (current != null) {
            result.add(current.value)
            current = current.next
        }
        return result
    }

    /**
     * Inserts [value] into the sorted position determined by [comparator].
     *
     * Assumes the list is already sorted. Scans from the head to locate the
     * correct insertion point – O(n) per call, making repeated calls O(n²).
     */
    fun insertSorted(value: T, comparator: Comparator<T>) {
        val newNode = Node(value)

        // Insert before head if the list is empty or value belongs at the front.
        if (head == null || comparator.compare(value, head!!.value) <= 0) {
            newNode.next = head
            head = newNode
            return
        }

        // Scan to find the node just before where value should land – O(n).
        var current = head!!
        while (current.next != null &&
            comparator.compare(value, current.next!!.value) > 0
        ) {
            current = current.next!!
        }
        newNode.next = current.next
        current.next = newNode
    }

    /**
     * Returns the number of elements.
     * O(n) – no cached counter is kept deliberately.
     */
    fun size(): Int {
        var count = 0
        var current = head
        while (current != null) {
            count++
            current = current.next
        }
        return count
    }

    /**
     * Removes all elements matching [predicate].
     * O(n).
     */
    fun removeIf(predicate: (T) -> Boolean) {
        // Skip matching nodes at the head.
        while (head != null && predicate(head!!.value)) {
            head = head!!.next
        }
        var current = head ?: return
        while (current.next != null) {
            if (predicate(current.next!!.value)) {
                current.next = current.next!!.next
            } else {
                current = current.next!!
            }
        }
    }

    /** Returns true when the list contains no elements. */
    fun isEmpty(): Boolean = head == null

    override fun toString(): String = toList().toString()
}
